leader-election-impl/src/main/java/com/palantir/paxos/VerifyingPaxosStateLog.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.paxos;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.function.Function;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.common.persist.Persistable;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.UnsafeArg;
+
+/**
+ * This implementation of {@link PaxosStateLog} uses one delegate as the source of truth, but also delegates all calls
+ * to the experimental delegate and verifies consistency.
+ *
+ * NOTE: while the read write lock guarantees atomicity in the absence of failures, the experimental log could still
+ * get out of sync with the source of truth if a write operation is performed on only one of the logs. Write operations
+ * have therefore been implemented to allow for simple re-hydration of the experimental log -- writes are performed on
+ * the current log first, while truncates are performed on the experimental log first.
+ */
+public final class VerifyingPaxosStateLog<V extends Persistable & Versionable> implements PaxosStateLog<V> {
+    private static final Logger log = LoggerFactory.getLogger(VerifyingPaxosStateLog.class);
+
+    private final ReadWriteLock lock = new ReentrantReadWriteLock(false);
+    private final PaxosStateLog<V> currentLog;
+    private final PaxosStateLog<V> experimentalLog;
+    private final Persistable.Hydrator<V> hydrator;
+
+    public VerifyingPaxosStateLog(Settings<V> settings) {
+        this.currentLog = settings.currentLog();
+        this.experimentalLog = settings.experimentalLog();
+        this.hydrator = settings.hydrator();
+    }
+
+    @Override
+    public void writeRound(long seq, V round) {
+        lock.writeLock().lock();
+        try {
+            currentLog.writeRound(seq, round);
+            experimentalLog.writeRound(seq, round);
+        } finally {
```

Comment:
This is a bit nasty, but I'd prefer

```
lock();
try {
  current.write(seq, round);
  try {
    experiment.write(seq, round);
  } catch (RuntimeException ex) {
    log.warn("I wrote to current but not experimental, please be careful with future errors for sequence number {}", SafeArg.of(seq), ex);
     // intentionally succeed, because we don't want to degrade service
  }
} finally {
  unlock();
}
```